### ISSUE TITLE
[FEATURE] Add universal dependency support

### ISSUE THREAD

**Issue by kkinder** [tiangolo/fastapi#688]
It can often be useful to essentially add an app-wide dependency. Suppose, for example, you want to enforce http basic auth on non-production environments:

```
async def basic_auth(credentials: HTTPBasicCredentials = Depends(security)):
    if (settings.SITEWIDE_BASIC_AUTH and (
            credentials.username != settings.SITEWIDE_BASIC_AUTH_USERNAME or
            credentials.password != settings.SITEWIDE_BASIC_AUTH_PASSWORD)):
        raise HTTPException(
            status_code=HTTP_401_UNAUTHORIZED,
            detail="Incorrect email or password",
            headers={"WWW-Authenticate": "Basic"},
        )

app.add_dependency(basic_auth)
```

Now, you could do something similar with middleware, but then you'd miss out on dependency injection. You can already do something similar for *included routers*, but not for the entire app.

Just a thought.

**Comment by prostomarkeloff** [comment]
LGTM. If i understood it right, `add_dependency` function simply will update all router routes dependencies?

**Comment by dmontagu** [comment]
You can achieve this for the full app now via:

1. `app_router = APIRouter()`
2. Add *all* desired routes to `app_router`
3. Call `app.include_router(app_router, dependencies=app_dependencies)`

Maybe you realized this already, just wanted to share in case it's useful (I use this pattern myself).

**Comment by kkinder** [comment]
Yeah, @dmontagu, that's more or less where I got the idea.

It's just sub-optimal because it requires that I factor out my use of `app` to `app_router` just to accommodate password-protecting the site, which seems inelegant.

**Comment by tiangolo** [comment]
Yeah, there's a couple of things that can be declared at router instantiation and at `.router_include`, bot others not. I agree it would be better to have consistency.

Being able to define dependencies and other stuff in `.router_include` is still valuable as it would allow re-using a router with different dependencies, tags, etc. under different paths, or by different apps. So I don't think removing those params from `.include_router` should be the way to go. But I agree it would make sense to have these things available for setting at instantiation.

I think the most difficult issue is handling defaults and overrides. Each point that can take parameters would have some default values or generate them at instantiation. That's normally handled with `is None`, but that works when there's a single place where a parameter could be declared. In our case, if a default is populated at some point then it's not possible to override that, at least not detectable with a simple `is None`, as it was already overwritten.

Nevertheless, I think I solved that (or something quite similar) in Typer. With some more complex extra tricks. Maybe we can replicate that here.

The other thing/issue is that currently, dependencies (and all those parameters) are actually stored at the `APIRoute` class level (for each function), not at the levels of the `APIRouter` and `FastAPI` classes. So, that would require quite some refactor, and tests.

I think it's worth it, but it requires quite some effort.

**Comment by daviskirk** [comment]
Python also has a topological sort builtin starting at 3.9: https://docs.python.org/3/library/graphlib.html?highlight=graph
which might be helpful when dealing with the logic of when which dependencies can/need to be run.

**Comment by dimaqq** [comment]
If I use `.include_router(router, dependencies=Depends(security))` and then one of the endpoints needs to see the credentials that `security` decoded from the `Authorization` header field, how do I go about that?

1. is there a way to poke at router dependency values somehow?
2. should I explicitly include that dependency in the special endpoint?
   * will fastapi eval it once and cache the result?
